packages/authorization-platformstate-writer/src/consumerServiceV1.ts

```diff
@@ -1,20 +1,512 @@
+              secondRetrievalStates.agreementEntry.state !==
+                agreementEntry.state ||
+              secondRetrievalStates.catalogEntry.state !== catalogEntry.state ||
+              secondRetrievalStates.purposeEntry.state !== purposeEntry.state
+            ) {
+              const updatedTokenClientPurposeEntry: TokenGenerationStatesClientPurposeEntry =
+                {
+                  ...clientPurposeEntry,
+                  GSIPK_consumerId_eserviceId: makeGSIPKConsumerIdEServiceId({
+                    consumerId: clientEntry.clientConsumerId,
+                    eserviceId:
+                      secondRetrievalStates.purposeEntry.purposeEserviceId,
+                  }),
+                  agreementId: extractAgreementIdFromAgreementPK(
+                    secondRetrievalStates.agreementEntry.PK
+                  ),
+                  agreementState: secondRetrievalStates.agreementEntry.state,
+                  GSIPK_eserviceId_descriptorId:
+                    makeGSIPKEServiceIdDescriptorId({
+                      eserviceId:
+                        secondRetrievalStates.purposeEntry.purposeEserviceId,
+                      descriptorId:
+                        secondRetrievalStates.agreementEntry
+                          .agreementDescriptorId,
+                    }),
+                  descriptorState: secondRetrievalStates.agreementEntry.state,
+                  descriptorAudience:
+                    secondRetrievalStates.catalogEntry.descriptorAudience,
+                  descriptorVoucherLifespan:
+                    secondRetrievalStates.catalogEntry
+                      .descriptorVoucherLifespan,
+                  purposeState: secondRetrievalStates.purposeEntry.state,
+                  purposeVersionId:
+                    secondRetrievalStates.purposeEntry.purposeVersionId,
+                  updatedAt: new Date().toISOString(),
+                };
+
+              await upsertTokenStateClientPurposeEntry(
+                updatedTokenClientPurposeEntry,
+                dynamoDBClient
+              );
+            }
+          }
+        }
+      }
+    })
+    .with({ type: "ClientPurposeRemoved" }, async (msg) => {
+      const clientId = unsafeBrandId<ClientId>(msg.data.clientId);
+      const purposeIdToRemove = unsafeBrandId<PurposeId>(msg.data.purposeId);
+
+      const pk = makePlatformStatesClientPK(clientId);
+      const clientEntry = await readClientEntry(pk, dynamoDBClient);
+
+      if (clientEntry) {
+        if (clientEntry.version > msg.version) {
+          return Promise.resolve();
+        } else {
+          const updatedPurposeIds = clientEntry.clientPurposesIds.filter(
+            (purposeId) => purposeId !== purposeIdToRemove
+          );
+          const GSIPK_clientId_purposeId = makeGSIPKClientIdPurposeId({
+            clientId,
+            purposeId: unsafeBrandId(msg.data.purposeId),
+          });
+
+          if (updatedPurposeIds.length > 0) {
```

Comment:
same considerations [here](https://github.com/pagopa/interop-be-monorepo/pull/1062#discussion_r1820599533)

Reply:
Updated